cupyx/scipy/sparse/linalg/_eigen.py

```diff
@@ -0,0 +1,158 @@
+import numpy
+import cupy
+
+
+def eigsh(a, k=6, which='LM', ncv=None, maxiter=None, tol=0,
+          return_eigenvectors=True):
+    """Finds ``k`` eigenvalues and eigenvectors of the real symmetric matrix.
+
+    Solves ``Ax = wx``, the standard eigenvalue problem for ``w`` eigenvalues
+    with corresponding eigenvectors ``x``.
+
+    Args:
+        a (cupy.ndarray or cupyx.scipy.sparse.csr_matrix): A symmetric square
+            matrix with dimension ``(n, n)``.
+        k (int): The number of eigenvalues and eigenvectors to compute. Must be
+            ``1 <= k < n``.
+        which (str): 'LM' or 'LA'. 'LM': finds ``k`` largest (in magnitude)
+            eigenvalues. 'LA': finds ``k`` largest (algebraic) eigenvalues.
+        ncv (int): The number of Lanczos vectors generated. Must be
+            ``k + 1 < ncv < n``. If ``None``, default value is used.
+        maxiter (int): Maximum number of Lanczos update iterations.
+            If ``None``, default value is used.
+        tol (float): Tolerance for residuals ``||Ax - wx||``. If ``0``, machine
+            precision is used.
+        return_eigenvectors (bool): If ``True``, returns eigenvectors in
+            addition to eigenvalues.
+
+    Returns:
+        tuple:
+            If ``return_eigenvectors is True``, it returns ``w`` and ``x``
+            where ``w`` is eigenvalues and ``x`` is eigenvectors. Otherwise,
+            it returns only ``w``.
+
+    .. seealso:: :func:`scipy.sparse.linalg.eigsh`
+
+    .. note::
+        This function uses the thick-restart Lanczos methos
+        (https://sdm.lbl.gov/~kewu/ps/trlan.html).
+
+    """
+    n = a.shape[0]
+    if a.ndim != 2 or a.shape[0] != a.shape[1]:
+        raise ValueError('expected square matrix (shape: {})'.format(a.shape))
+    if a.dtype.char not in 'fdFD':
+        raise TypeError('unsupprted dtype (actual: {})'.format(a.dtype))
+    if k <= 0:
+        raise ValueError('k must be greater than 0 (actual: {})'.format(k))
+    if k >= n:
+        raise ValueError('k must be smaller than n (actual: {})'.format(k))
+    if which not in ('LM', 'LA'):
+        raise ValueError('which must be \'LM\' or \'LA\' (actual: {})'
+                         ''.format(which))
+    if ncv is None:
+        ncv = min(max(8 * k, 16), n - 1)
+    else:
+        ncv = min(max(ncv, k + 2), n - 1)
```

Comment:
I'm sure that individual values are not significant, but the expressions look different from that on SciPy's reference. Where do they come from?
> it is recommended that ncv > 2*k. Default: `min(n, max(2*k + 1, 20))`

Reply:
The larger default value of ncv (the number of Lanczos vectors) compared to SciPy is based on experimental results, adjusted to roughly match SciPy's calculation results and the residuals (||Ax - wx||).
ARPACK, the library used by SciPy, uses "implicitly restarted Lanczos", but this implementation uses "thick restart Lanczos", which is a simpler restarting scheme. According to https://sdm.lbl.gov/~kewu/ps/trlan.html, thick restart is equivalent to implicit restart, but from the experimental results, it seems that the number of Lanczos vectors needs to be increased.

Reply:
Thanks!